Add warning for UO execution schedule, fix sibling transfer execution schedule

GiudGiud opened the pull request:
refs https://github.com/idaholab/moose/issues/31074

GiudGiud committed:
Better checks/parameters on discontinous sampling in sampler classes

GiudGiud committed:
Add exec transfer to all user objects
Note:
AuxKernels are executed before transfers on timestep_end but not on timestep_begin
So we could use the same thing for auxkernels on timestep_begin to be able to send updated data from auxkernels

On the other hand, UOs are executed after transfers on both timestep begin and timestep end,
so that's why we need the option to be able to execute on TRANSFER
refs #31074

GiudGiud committed:
Do not execute sibling transfers all the time

GiudGiud committed:
Add a warning on known-to-be-bad schedule for transferred UO execution
Add a boolean to allow disabling the warning

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3092816/), step Docs: sync website on 63174b4 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31106/site)

This comment will be updated on new commits.

GiudGiud committed:
Execute between_multiapp transfers on the target app schedule rather than the source app

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3099080/), step Generate coverage on 5241e44 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ee92e3</th>
      <th colspan=4>#31106 5241e4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.54%</td>
      <td>85.56%</td>
      <td>+0.01%</td>
      <td>76.92%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>116194</td>
      <td>116240</td>
      <td>+46</td>
      <td>30</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19634</td>
      <td>19620</td>
      <td>-14</td>
      <td>9</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31106/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31106/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>ee92e3</th>
      <th colspan=4>#31106 5241e4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>86.01%</td>
      <td>86.00%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>8180</td>
      <td>8179</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1331</td>
      <td>1332</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31106/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31106/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31106/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 76.92% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5609,10 +5609,12 @@ FEProblemBase::addTransfer(const std::string & transfer_name,
       multiapp = getMultiApp(parameters.get<MultiAppName>("multi_app"));
-    else if (parameters.isParamValid("from_multi_app"))
-      multiapp = getMultiApp(parameters.get<MultiAppName>("from_multi_app"));
+    // This catches the sibling transfer case, where we want to be executing only as often as the
+    // receiving application. A transfer 'to' a multiapp is executed before that multiapp
     else if (parameters.isParamValid("to_multi_app"))
       multiapp = getMultiApp(parameters.get<MultiAppName>("to_multi_app"));
+    else if (parameters.isParamValid("from_multi_app"))
+      multiapp = getMultiApp(parameters.get<MultiAppName>("from_multi_app"));
```
you could reduce the diff here

GiudGiud replied:
not sure I understand what you mean

lindsayad replied:
I see

lindsayad commented on framework/src/transfers/MultiAppPostprocessorTransfer.C:
```diff
@@ -38,6 +38,12 @@ MultiAppPostprocessorTransfer::validParams()
                              reduction_type,
                              "The type of reduction to perform to reduce postprocessor "
                              "values from multiple SubApps to a single value");
+  params.addParam<bool>(
+      "warn_source_object_execution_schedule",
+      true,
+      "Emit a warning when the transfer execution schedule is detected to lag "
+      "information from the postprocessor. Note that the check cannot detect all "
+      "potential wrong combinations of postprocessor/transfer execution schedules");
```
I wonder if you could make a static method on the `MultiAppTransfer` that takes a reference to `InputParameters` and adds this parameter and call it from these derived classes. Then if you ever want to change the doc string or the default you could just do it one place

GiudGiud replied:
I think I can do it

GiudGiud replied:
done

GiudGiud committed:
Address Alex's review
Place the UO execution check parameter for triggering the warning in a static so it is shared by
all 5 UO-related or PP-related transfers
